### ISSUE TITLE
[Feature] Add emoji reactions for all message types

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#14180]
<Details>
<Summary>Original P/S and project checklist</Summary>
## Proposal
Add Emoji Reactions for All Message Types

*Problem*
Text chat is very informational, but sometimes you want to participate in a way that is more emotional than informational -- to show support, comfort, disagreement, etc.  In this case, a full-text message feels overkill, meaning oftentimes there is a desire to casually and emotionally participate that doesn’t clear the bar for a full-text message, and thus results in deciding not to participate at all.
Also, joining chats comes with an opportunity cost. You may get many notifications that distract you or you may get roped into conversations that you aren’t as passionate about. If you comment and someone replies, there’s a heightened expectation that you should reply back to them in a reasonable timeframe.

*Solution*
Let’s implement reactions into NewDot so users can react with emojis to any message type. Reactions became a standard in chatting apps as they offer a quick way to respond to messages and bring an additional layer of communication. They are also less distracting, as we don’t send notifications for reactions.

*Design Doc* https://docs.google.com/document/d/14VKy0ozcY_YVEmw0wV2ri_OBi7XNGUFVHhK3ei1MNWY/edit#heading=h.ns12dr1f35rx

## Tasks

- [x] Post **Proposal (full Problem/Solution statement)** in `#expensify-open-source` ([example](https://expensify.slack.com/archives/C01GTK53T8Q/p1673641691365339))
- [x] Wait at least one full business day, and until the post has a majority (2/3) of positive reactions (👍)
- [x] Paste Proposal in the space above with a link to the Slack thread
- [x] Email `strategy@expensify.com` and paste in the Proposal
- [x] Fill out the **High-level overview of the problem**, **Timeline**, and **Terminology** sections of the Design Doc
- [x] Email `strategy@expensify.com` (continue the same email chain as before) with the link to your Design Doc
- [x] Host a **pre-design** meeting ([example](https://expensify.slack.com/archives/C01GTK53T8Q/p1665794669539419)) in `#expensify-open-source` to discuss any necessary details in public before filling out the **High-level of proposed solution** section.
- [x] Fill out the **High-level of proposed solution** section
- [x] Email `stategy@expensify.com` again with links to the doc and pre-design conversation in Slack
- [x] Add the `DesignDocReview` label to get the **High-level of proposed solution** section reviewed
- [x] Respond to any questions or concerns and bring up blockers in Slack to get a consensus if necessary
- [x] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [x] Host another pre-design meeting in `#expensify-open-source` to ask for engineering feedback on the technical solution.
- [x] Fill out the **Detailed implementation of the solution** and related sections.
- [x] Re-add the `DesignDocReview` label to this issue
- [x] Respond to any questions or concerns and bring up blockers in Slack to get consensus if necessary
- [x] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [x] Email `strategy@expensify.com` one last time to let them know the Design Doc is moving into the implementation phase
- [ ] Implement the changes
- [ ] Send out a follow up email to `strategy@expensify.com` once everything has been implemented and do a **Project Wrap-Up** retrospective that provides:
    - Summary of what we accomplished with this project
    - What went well?
    - What could we have done better?
    - What did we learn?
</Details>

GH Tracking:
- [x] https://github.com/Expensify/App/issues/15054 - @stitesExpensify 
- [x] https://github.com/Expensify/App/issues/15056 - @stitesExpensify
- [x] https://github.com/Expensify/App/issues/15061 - @stitesExpensify
- [x] https://github.com/Expensify/App/issues/15063 - @stitesExpensify
- [x] https://github.com/Expensify/App/issues/15062 - @stitesExpensify
- [x] https://github.com/Expensify/App/issues/15125 - @hannojg 
- [x] https://github.com/Expensify/App/issues/15126 - @hannojg 
- [x] https://github.com/Expensify/App/issues/15127 - @hannojg 
- [x] https://github.com/Expensify/App/issues/15128 - @hannojg 

Cleanup Issues:
- [x] https://github.com/Expensify/App/issues/15786 - @stitesExpensify 
- [x] https://github.com/Expensify/App/issues/15787 - @stitesExpensify 
- [x] https://github.com/Expensify/App/issues/16549 - @stitesExpensify 
- [ ] https://github.com/Expensify/App/issues/16506 - @stitesExpensify

**Comment by JmillsExpensify** [comment]
Copying over the pre-design notes from our in-person meeting in San Franciso.

**Emoji Reactions Pre-design**
 - Emoji is added via the comment action menu
   - Hover - option to add an emoji via the picker
   - Right-click on web - potentially showing the most-used emojis at the top above options
   - Long-press on mobile - Same as right click on web
 - Emoji reactions appear immediately below the message.
 - If emoji reactions already exist, add an emoji button inline.
 - Emojis are shown in the order that they are added.
 - Clicking on an emoji reaction adds it. Clicking on one you added removes it
 - We’ll use Pattern A for reactions on comments
 - If you react to certain emojis, those will appear blue/border
 - To see who reacted to an emoji…
   - Long-press - who selected the emoji is visible in the drawer, grouped by person sorted by time
   - Right click - Tooltip for that emoji, with the list of usernames that selected that emoji (fallback to phone/email)
   - Hover - Tooltip for that emoji, with the list of usernames that selected that emoji (fallback to phone/email)
 - We always list the person’s name that selected the emoji
 - For mobile, the height of the bottom-dock will be dynamic based on it’s comment
 - For mobile, the drawer will support swiping
 
 **Items that came up in discussion**
 - What about emoji reaction limits? We won’t worry about this for now.
 - Should we notify on reactions? No (though we could change this in the future)
 - Can you react on every type of reportAction? Yes
 - Can you react to the welcome message? No
 - Should the chat become unread if you receive a reaction? No
 - Is a reaction a message (like WhatsApp or iMessage)? No

**Comment by JmillsExpensify** [comment]
Went ahead and added myself as the BZ helping push this forward.

**Comment by Szymon20000** [comment]
Sure!

**Comment by hannojg** [comment]
@puneetlath Szymon and I are sharing the work on this one. Is it possible to assign me as well? Thanks 😊

**Comment by stitesExpensify** [comment]
@Szymon20000 @hannojg FYI I just updated the OP to match our new Design Doc GH Issue template. It provides an outline of our internal process, and should help avoid any confusion that may be caused by the design doc itself. If you have any questions, obviously feel free to reach out to me or in #expensify-margelo!

**Comment by hannojg** [comment]
Following the official procedure I posted a problem solution statement [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1674070038007679)

**Comment by hannojg** [comment]
Can we please update the issue with the P/S Statement: https://expensify.slack.com/archives/C01GTK53T8Q/p1674214838372159?thread_ts=1674070038.007679&cid=C01GTK53T8Q

Thanks! 😊

**Comment by stitesExpensify** [comment]
Done!

**Comment by hannojg** [comment]
It also seems like I am unable to mail `strategy@expensify.com`? Can we whitelist the `@margelo.io` emails, or do you guys want to send out the emails?

**Comment by stitesExpensify** [comment]
@hannojg I just confirmed that it is an internal only list. For now so that we can keep this moving I'll send it out for you while we get that changed.

If you want to post the contents of your email here, I will take care of it ASAP 😄

**Comment by hannojg** [comment]
Yes, that would be great, thx!

**Title:** [Feature proposal] Add Emoji Reactions for All Message Types

Problem
Text chat is very informational, but sometimes you want to participate in a way that is more emotional than informational -- to show support, comfort, disagreement, etc. In this case, a full-text message feels overkill, meaning oftentimes there is a desire to casually and emotionally participate that doesn’t clear the bar for a full-text message, and thus results in deciding not to participate at all.
Also, joining chats comes with an opportunity cost. You may get many notifications that distract you or you may get roped into conversations that you aren’t as passionate about. If you comment and someone replies, there’s a heightened expectation that you should reply back to them in a reasonable timeframe.

Solution
Let’s implement reactions into NewDot so users can react with emojis to any message type. Reactions became a standard in chatting apps as they offer a quick way to respond to messages and bring an additional layer of communication. They are also less distracting, as we don’t send notifications for reactions.

Slack thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1674214838372159?thread_ts=1674070038.007679&cid=C01GTK53T8Q

**Comment by stitesExpensify** [comment]
Strategy@ email was sent out today! We should be ready for the high level review on Monday

**Comment by Gonals** [comment]
I have read and reviewed this Design Doc!

**Comment by stitesExpensify** [comment]
High level design is ready for review!

**Comment by tgolen** [comment]
I have read and reviewed this Design Doc!

**Comment by luacmartins** [comment]
I have read and reviewed this Design Doc!

**Comment by thienlnam** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
I have read and reviewed this Design Doc!

**Comment by jliexpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by hannojg** [comment]
@stitesExpensify @JmillsExpensify @roryabraham We updated the design doc and went through all the comments. We are still missing enough approvals from the reviews I'd say. How can we request a new review of the changes of the document? 😊

**Comment by mountiny** [comment]
> How can we request a new review of the changes to the document (the bot removed the label)? 😊

All the review issues created are made Weekly, we can make them Daily so people act with higher urgency so you dont have to wait for this. The label is designed so once everyone who got assigned reviews, it should be enough reviews. 

The people who already reviewed will most likely come check in on the changes and review again. Otherwise we just gotta wait a bit as people have obviously other stuff to do 😅

**Comment by robertjchen** [comment]
I have read and reviewed this Design Doc!

**Comment by peterdbarkerUK** [comment]
I have read and reviewed this Design Doc!

**Comment by justinpersaud** [comment]
I have read and reviewed this Design Doc!

**Comment by francoisl** [comment]
I have read and reviewed this Design Doc!

**Comment by flodnv** [comment]
I have read and reviewed this Design Doc!

**Comment by mountiny** [comment]
I have read and reviewed this Design Doc!

**Comment by puneetlath** [comment]
I have read and reviewed this Design Doc!

**Comment by trjExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by quinthar** [comment]
I have read and reviewed this Design Doc!

**Comment by RobertLadue** [comment]
I have read and reviewed this Design Doc!

**Comment by jeremyhollmexpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by hayata-suenaga** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
I have read and reviewed this Design Doc!

**Comment by AndrewGable** [comment]
I have read and reviewed this Design Doc!

**Comment by puneetlath** [comment]
I have read and reviewed this Design Doc!

**Comment by iwiznia** [comment]
I have read and reviewed this Design Doc!

**Comment by dangrous** [comment]
I have read and reviewed this Design Doc!

**Comment by trjExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by robertjchen** [comment]
I have read and reviewed this Design Doc!

**Comment by mountiny** [comment]
I have read and reviewed this Design Doc!

**Comment by francoisl** [comment]
I have read and reviewed this Design Doc!

**Comment by tgolen** [comment]
I have read and reviewed this Design Doc!

**Comment by RobertLadue** [comment]
I have read and reviewed this Design Doc!

**Comment by JmillsExpensify** [comment]
I have read and reviewed this design doc!

**Comment by stitesExpensify** [comment]
All backend PRs are in review! FrontEnd PRs should be coming out today!

**Comment by JmillsExpensify** [comment]
Wow! Well look at that 🚀

**Comment by cdraeger11** [comment]
I have read and reviewed this design doc!

**Comment by JmillsExpensify** [comment]
Detailed implementation is on-going.

**Comment by stitesExpensify** [comment]
Web and Auth are merged and deployed! /App MVP is in final review!

**Comment by stitesExpensify** [comment]
MVP should be merged today or first thing tomorrow morning, with the goal being to deploy it before ExpensiconX begins!

One thing to note is that we found a bug today in how we are returning reportActions with `type == IOU` which causes them to disappear on the front end. To make sure that the MVP still goes out the door in time, we will be temporarily disabling the ability to react to IOUs, and enabling it once the backend fix has been deployed.

**Comment by aldo-expensify** [comment]
Found a little bug I think: https://github.com/Expensify/App/issues/15786

I was so surprise to see today that we had reactions! this was very quickly implemented 🚀

**Comment by stitesExpensify** [comment]
Thanks for making that issue! We've been chatting about it in slack [here](https://expensify.slack.com/archives/C049HHMV9SM/p1678372751210589) but I'm glad to have it formally written down 😄

**Comment by stitesExpensify** [comment]
This is going very well! The only original issue that we have remaining is for seeing who reacted on mobile.

We tested reactions out a lot during ECX and found some issues that we need to clean up. I have updated the OP to include them

**Comment by alex-mechler** [comment]
We found another bug, where some emojis have their color overwritten by the `color` style. Is there any reason we have this set for emojis? https://github.com/Expensify/App/issues/16018

**Comment by JmillsExpensify** [comment]
Just a couple of issues still outstanding.

**Comment by JmillsExpensify** [comment]
> I am working on emojis disappearing when reacting rapidly this week

Funny enough I just hit this one testing reactions on IOUs.

**Comment by JmillsExpensify** [comment]
It looks like we're down to one final issue before we can close this one out.

**Comment by stitesExpensify** [comment]
Tim is still working on the fast adding/deleting, and I will be taking over again once he leaves for his sabbatical

**Comment by stitesExpensify** [comment]
We are actively deploying PRs to finish fixing the reaction flickering issue. We had to revert one PR last week which slowed us down, but we are marching forward!

**Comment by JmillsExpensify** [comment]
Looks like we are getting close to that last issue still. See the linked PRs for more.

**Comment by stitesExpensify** [comment]
I'm still working on the final issue. The merge conflicts are off the charts on the original PR that @tgolen made. With all of the refactors etc I've decided to start a new PR and pull the changes over manually since so much has changed. Top priority next week.

**Comment by stitesExpensify** [comment]
I fixed all of the conflicts, but we found another [breaking issue](https://github.com/Expensify/Expensify/issues/298245) with regards to how we save skintones. I created a new issue and have a PR up for the fix, but it does mean we need to wait for another couple of deploys before we can move forward with the /app PR again.

**Comment by stitesExpensify** [comment]
We're almost done, waiting for one last review

**Comment by stitesExpensify** [comment]
Reactions are no longer disappearing when adding/removing quickly! We are finishing up some cleanup and then is finally finished!

**Comment by JmillsExpensify** [comment]
Woo! Great work all around.

**Comment by stitesExpensify** [comment]
Still cleaning up

**Comment by stitesExpensify** [comment]
I think that we can close this. We are still cleaning up, but the feature is now complete and fully functional! I will continue to clean up the code as I have time, but this project is done. I'll send out an email soon!

**Comment by JmillsExpensify** [comment]
I'm with you. Let's close it out!
